base/fastmath.jl

```diff
@@ -0,0 +1,326 @@
+abs2_fast{T<:ComplexTypes}(x::T) = real(x)*real(x) + imag(x)*imag(x)
+add_fast{T<:ComplexTypes}(x::T, y::T) = T(real(x)+real(y), imag(x)+imag(y))
+conj_fast{T<:ComplexTypes}(x::T) = T(real(x), -imag(x))
+function div_fast{T<:ComplexTypes}(x::T, y::T)
+    T(real(x)*real(y) + imag(x)*imag(y),
+       imag(x)*real(y) - real(x)*imag(y)) / abs2(y)
+end
+inv_fast{T<:ComplexTypes}(x::T) = conj(x) / abs2(x)
+sign_fast{T<:ComplexTypes}(x::T) = x / abs(x)
+sub_fast{T<:ComplexTypes}(x::T, y::T) = T(real(x)-real(y), imag(x)-imag(y))
+function mul_fast{T<:ComplexTypes}(x::T, y::T)
+    T(real(x)*real(y) - imag(x)*imag(y),
+       real(x)*imag(y) + imag(x)*real(y))
+end
+
+eq_fast{T<:ComplexTypes}(x::T, y::T) = (real(x)==real(y)) & (imag(x)==imag(y))
+ne_fast{T<:ComplexTypes}(x::T, y::T) = !(x==y)
+
+# fall-back implementations and type promotion
+
+for op in (:+, :-, :*, :/, :(==), :!=, :<, :<=,
+           :abs, :abs2, :cmp, :conj, :inv, :mod, :rem, :sign)
+    op_fast = fast_op[op]
+    @eval begin
+        # fall-back implementation for non-numeric types
+        ($op_fast)(xs...) = ($op)(xs...)
+        # type promotion
+        ($op_fast)(x::Number, y::Number, zs::Number...) =
+            ($op_fast)(promote(x,y,zs...)...)
+        # fall-back implementation that applies after promotion
+        ($op_fast){T<:Number}(x::T,ys::T...) = ($op)(x,ys...)
+    end
+end
+
+
+
+# Math functions
+
+# builtins
+
+pow_fast{T<:FloatTypes}(x::T, y::Integer) =
+    box(T, Base.powi_llvm(unbox(T,x), unbox(Int32,int32(y))))
+
+# TODO: Change sqrt_llvm intrinsic to avoid nan checking; add nan
+# checking to sqrt in math.jl; remove sqrt_llvm_fast intrinsic
+sqrt_fast{T<:FloatTypes}(x::T) = box(T, Base.sqrt_llvm_fast(unbox(T,x)))
+
+# libm
+
+const libm = Base.libm_name
+
+for f in (:acos, :acosh, :asin, :asinh, :atan, :atanh, :cbrt, :cos,
+          :cosh, :exp2, :exp, :expm1, :lgamma, :log10, :log1p, :log2,
+          :log, :sin, :sinh, :tan, :tanh)
+    f_fast = fast_op[f]
+    @eval begin
+        ($f_fast)(x::Float32) =
+            ccall(($(string(f,"f")),libm), Float32, (Float32,), x)
+        ($f_fast)(x::Float64) =
+            ccall(($(string(f)),libm), Float64, (Float64,), x)
+    end
+end
+
+
+for f in (:^, :atan2)
+    f_fast = fast_op[f]
+    if f==:^; f=:pow; end       # special case
+    @eval begin
+        ($f_fast)(x::Float32, y::Float32) =
+            ccall(($(string(f,"f")),libm), Float32, (Float32,Float32), x, y)
+        ($f_fast)(x::Float64, y::Float64) =
+            ccall(($(string(f)),libm), Float64, (Float64,Float64), x, y)
+    end
+end
+
+pow_fast(x::Float32, y::Float32) =
+    ccall(("powf",libm), Float32, (Float32,Float32), x, y)
+pow_fast(x::Float64, y::Float64) =
+    ccall(("pow",libm), Float64, (Float64,Float64), x, y)
+
+atan2(x::Float32, y::Float32) =
+    ccall(("atan2f",libm), Float32, (Float32,Float32), x, y)
+atan2(x::Float64, y::Float64) =
+    ccall(("atan2",libm), Float64, (Float64,Float64), x, y)
+
```

Comment:
Should these be `atan2_fast`?  Also, I think the method generating `for` loop defined above can now be deleted.